### PR TITLE
SymIntify randint and randperm

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98968
* #99157

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

**Review by xw285cornell — APPROVED** [review]
Thanks!!

**Comment by xw285cornell** [comment]
Looks it needs to be a bit more involved: 

aten/src/ATen/functorch/BatchRulesRandomness.cpp:319:12: error: no matching function for call to 'random_batching_rule'
    return random_batching_rule<decltype(&rand_int_low_wrapper<F, Func, T0, T1, T...>),
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
aten/src/ATen/functorch/BatchRulesRandomness.cpp:466:3: note: in instantiation of member function 'at::functorch::RandTwoLeadingScalarsBatchRuleHelper<at::Tensor (*)(c10::SymInt, c10::SymInt, c10::ArrayRef<c10::SymInt>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>), &at::_ops::randint_low::call, c10::guts::typelist::typelist<c10::SymInt, c10::SymInt, c10::ArrayRef<c10::SymInt>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>>>::apply' requested here
  RAND_TWO_LEADING_SCALARS_BATCH_RULE(randint, low);
  ^
aten/src/ATen/functorch/BatchRulesRandomness.cpp:417:113: note: expanded from macro 'RAND_TWO_LEADING_SCALARS_BATCH_RULE'
                                c10::guts::function_traits<decltype(ATEN_FN2(op, overload))>::parameter_types>::apply))
                                                                                                                ^
aten/src/ATen/functorch/BatchRulesRandomness.cpp:23:8: note: candidate function template not viable: no known conversion from 'c10::SymInt' to 'long' for 2nd argument
Tensor random_batching_rule(SymIntArrayRef shape, ExtraArgs... extra_args) {
       ^
aten/src/ATen/functorch/BatchRulesRandomness.cpp:319:12: error: no matching function for call to 'random_batching_rule'
    return random_batching_rule<decltype(&rand_int_low_wrapper<F, Func, T0, T1, T...>),
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
aten/src/ATen/functorch/BatchRulesRandomness.cpp:467:3: note: in instantiation of member function 'at::functorch::RandTwoLeadingScalarsBatchRuleHelper<at::Tensor (*)(c10::SymInt, c10::SymInt, c10::ArrayRef<c10::SymInt>, c10::optional<at::Generator>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>), &at::_ops::randint_low_generator::call, c10::guts::typelist::typelist<c10::SymInt, c10::SymInt, c10::ArrayRef<c10::SymInt>, c10::optional<at::Generator>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>>>::apply' requested here
  RAND_TWO_LEADING_SCALARS_BATCH_RULE(randint, low_generator);
  ^
aten/src/ATen/functorch/BatchRulesRandomness.cpp:417:113: note: expanded from macro 'RAND_TWO_LEADING_SCALARS_BATCH_RULE'
                                c10::guts::function_traits<decltype(ATEN_FN2(op, overload))>::parameter_types>::apply))
                                                                                                                ^
aten/src/ATen/functorch/BatchRulesRandomness.cpp:23:8: note: candidate function template not viable: no known conversion from 'c10::SymInt' to 'long' for 2nd argument
Tensor random_batching_rule(SymIntArrayRef shape, ExtraArgs... extra_args) {
       ^

**Comment by ezyang** [comment]
I think I've nailed the functorch, still building

File: aten/src/ATen/functorch/BatchRulesRandomness.cpp [hunk 1]
```diff
@@ -279,7 +279,7 @@ struct RandomBatchRuleHelper<F, Func, typelist<T1, T...>> {
 };
 
 template <typename F, F Func, typename... T>
-Tensor rand_int_wrapper(SymIntArrayRef shape, int64_t high, T... extra_args) {
+Tensor rand_int_wrapper(SymIntArrayRef shape, c10::SymInt high, T... extra_args) {
```
Comment (xw285cornell) [inline aten/src/ATen/functorch/BatchRulesRandomness.cpp:282]: nit: probably ok w/o c10 namespace

File: aten/src/ATen/functorch/BatchRulesRandomness.cpp [hunk 2]
```diff
@@ -298,10 +298,10 @@ struct RandIntBatchRuleHelper;
 
 template <typename F, F Func, typename T1, typename T2, typename... T>
 struct RandIntBatchRuleHelper<F, Func, typelist<T1, T2, T...>> {
-  static Tensor apply(int64_t high, SymIntArrayRef shape, T... extra_args) {
+  static Tensor apply(c10::SymInt high, SymIntArrayRef shape, T... extra_args) {
     return random_batching_rule<decltype(&rand_int_wrapper<F, Func, T...>),
                                 &rand_int_wrapper<F, Func, T...>,
-                                int64_t, T...>(shape, high, std::forward<T>(extra_args)...);
+                                c10::SymInt, T...>(shape, high, std::forward<T>(extra_args)...);
```
Comment (Skylion007) [inline aten/src/ATen/functorch/BatchRulesRandomness.cpp:304]:
```suggestion
                                c10::SymInt, T...>(std::move(shape), high, std::forward<T>(extra_args)...);

File: aten/src/ATen/functorch/BatchRulesRandomness.cpp [hunk 3]
```diff
@@ -279,7 +279,7 @@ struct RandomBatchRuleHelper<F, Func, typelist<T1, T...>> {
 };
 
 template <typename F, F Func, typename... T>
-Tensor rand_int_wrapper(SymIntArrayRef shape, int64_t high, T... extra_args) {
+Tensor rand_int_wrapper(SymIntArrayRef shape, c10::SymInt high, T... extra_args) {
   return Func(high, shape, std::forward<T>(extra_args)...);
```
Comment (Skylion007) [inline aten/src/ATen/functorch/BatchRulesRandomness.cpp:283]:
```suggestion
  return Func(std::move(high), shape, std::forward<T>(extra_args)...);
```

**Review by albanD — COMMENTED** [review]
Sounds good but tests need updating.

**Comment by ezyang** [comment]
So the reason tests are failing are because pass large negative numbers to randint, e.g., -9223372036854775808, which aren't representable in SymInt. Need to come up with a strategy for this.

**Pull request closed without merging**